docs: Implement style guide rules - active voice, misc #2 (#23273)
MystenLabs/sui: docs/content/concepts/sui-move-concepts/conventions.mdx

description: Recommended Move 2024 best practices for Sui development.
---

This guide outlines recommended conventions and best practices for writing Move smart contracts on Sui. Following these guidelines helps create maintainable, secure, and composable code that aligns with ecosystem standards.

These conventions are recommendations, not strict rules. They represent patterns that have proven effective across many Sui projects and help ensure consistency across the ecosystem.

## Organization principles

> ### Package

Author: ```suggestion
```

A Sui package consists of:

> - a `sources` directory containing the Move code uploaded on-chain  
> - a `Move.toml` manifest file that declares dependencies and package metadata  
> - a `Move.lock` file that the toolchain generates to lock dependency versions and track published and upgraded package addresses across networks  

Author: ```suggestion
- A `sources` directory containing the Move code uploaded on-chain.
- A `Move.toml` manifest file that declares dependencies and package metadata.
- A `Move.lock` file that the toolchain generates to lock dependency versions and track published and upgraded package addresses across networks.
```

Always commit the `Move.lock` file (do not add it to `.gitignore`). Use [automated address management](https://docs.sui.io/concepts/sui-move-concepts/packages/automated-address-management) instead of the old `published-at` field in the manifest.

> Optional directories:
> 
> - `tests` for package tests  
> - `examples` for usage examples  
> 
> These are not uploaded when you publish.

Author: ```suggestion
Optionally, you can add a `tests` directory to contain the tests for the package and an `examples` directory to provide use cases for the package. Neither directory is uploaded on chain when you publish the package.
```

## Naming conventions

> 

Author: ```suggestion

Adhering to naming conventions in your code helps readability and ultimately makes your codebase easier to maintain. The following sections outline the key naming conventions to follow when writing Move code.

```

### Constants

> - Constants: uppercase snake case.  

Author: ```suggestion
Constants should be uppercase and formatted as snake case. Errors are specific constants that use PascalCase and start with an E. Make them descriptive.
```

> - Errors: PascalCase starting with `E`.  

Author: ```suggestion
```

```move
module conventions::constants;

const MAX_NAME_LENGTH: u64 = 64;
const EInvalidName: u64 = 0; // correct
const E_INVALID_NAME: u64 = 0; // wrong
```

### Structs

- Declare abilities in this order: `key`, `copy`, `drop`, `store`.  
- Avoid using `Potato` as a suffix (this signals missing abilities).  
- Use positional fields for simple wrappers, dynamic field keys, or tuples.  
- Use the `Event` suffix for event structs.  

```move
module conventions::request;

public struct ReceiptKey(ID) has copy, drop, store;

> public struct Receipt<Data> has key, store {

Author: ```suggestion
// dynamic field
public struct Receipt<Data> has key, store {
```

    id: UID,
    data: Data
}

public struct Request(); // correct
public struct RequestPotato {} // wrong
```

### CRUD function names

Standard names:

- `new` → create an empty object  
- `empty` → create an empty struct  
- `create` → create an initialized object  
- `add` / `remove` → add or remove a value  
- `exists` / `contains` → check existence  
- `borrow` / `borrow_mut` → return a reference  
- `property_name` / `property_name_mut` → return a field  
- `drop` → drop a struct  
- `destroy` / `destroy_empty` → destroy an object  
- `to_name` / `from_name` → type transformations  

### Generics

Use either single letters (`T`, `U`) or descriptive names. Prioritize readability.

> ```move

Author: ````suggestion
```move
module conventions::generics;
// single letter name
````
